### PR TITLE
8343941: IGV: dump graph at different register allocation steps

### PR DESCRIPTION
**Pull request by robcasloz**
This changeset dumps C2's low-level intermediate representation at the following intermediate register allocation points:
- Initial liveness: after initial liveness information is computed.
- Aggressive coalescing: after aggressively coalescing live ranges and destructing SSA.
- Initial spilling: after initial round of spilling derived from physical interference graph construction.
- Conservative coalescing: after each round of conservative (colorability-preserving) coalescing (if `OptoCoalesce` is enabled).
- Iterative spilling: after each round of spilling.
- After iterative spilling: after the main register allocation loop.
- Post-allocation copy removal: after peephole copy removal.
- Merge multiple definitions: after local merging of equivalent nodes related by the same live range.
- Fix up spills: convert load-store spills into memory operand accesses ("CISC spilling") if allowed by the target platform and `UseCISCSpill` is enabled.

The new dumps have already proved to be useful in the investigation of [JDK-8331295](https://bugs.openjdk.org/browse/JDK-8331295).

#### Testing

- tier1-3 (windows-x64, linux-x64, linux-aarch64, macosx-x64, and macosx-aarch64).
- Tested automatically that dumping, scheduling, and viewing hundreds of the new graphs does not trigger any failure on HotSpot or IGV (by instrumenting IGV to schedule and view graphs eagerly and running `java -Xbatch -XX:PrintIdealGraphLevel=4`).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343941](https://bugs.openjdk.org/browse/JDK-8343941): IGV: dump graph at different register allocation steps (**Enhancement** - P4)

### Reviewers
 * [Daniel Lundén](https://openjdk.org/census#dlunden) (@dlunde - Committer) Review applies to [e44fa796](https://git.openjdk.org/jdk/pull/22017/files/e44fa796ddd385f853e0ab674c558ab6c4ac3feb)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Damon Fenacci](https://openjdk.org/census#dfenacci) (@dafedafe - Committer)

### Reviewers without OpenJDK IDs
 * @abdelhak-zaaim (no known openjdk.org user name / role) Review applies to [e44fa796](https://git.openjdk.org/jdk/pull/22017/files/e44fa796ddd385f853e0ab674c558ab6c4ac3feb)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22017/head:pull/22017` \
`$ git checkout pull/22017`

Update a local copy of the PR: \
`$ git checkout pull/22017` \
`$ git pull https://git.openjdk.org/jdk.git pull/22017/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22017`

View PR using the GUI difftool: \
`$ git pr show -t 22017`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22017.diff">https://git.openjdk.org/jdk/pull/22017.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22017#issuecomment-2469749842)
</details>

### CODE REVIEW

**Comment by dlunde** [comment]
Looks great, and I can confirm the new phases are very useful!

**Comment by robcasloz** [comment]
> Looks great, and I can confirm the new phases are very useful!

Thanks Daniel, feel free to review!

**Review by dlunde — APPROVED** [review]

File: src/hotspot/share/opto/phasetype.hpp [hunk 1]
```diff
@@ -93,6 +93,15 @@
...
   flags(GLOBAL_CODE_MOTION,             "Global code motion") \
+  flags(INITIAL_LIVENESS,               "Initial liveness") \
+  flags(AGGRESSIVE_COALESCING,          "Aggressive coalescing") \
+  flags(INITIAL_SPILLING,               "Initial spilling") \
+  flags(CONSERVATIVE_COALESCING,        "Conservative coalescing") \
+  flags(ITERATIVE_SPILLING,             "Iterative spilling") \
+  flags(AFTER_ITERATIVE_SPILLING,       "After iterative spilling") \
+  flags(POST_ALLOCATION_COPY_REMOVAL,   "Post-allocation copy removal") \
+  flags(MERGE_MULTIDEFS,                "Merge multiple definitions") \
+  flags(FIXUP_SPILLS,                   "Fix up spills") \
```
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/hotspot/share/opto/phasetype.hpp]: Thanks, done in commit e44fa796.

**Comment by robcasloz** [comment]
> Just an idea, since you've provided a nice description for each phase in the PR description, should we add these in phasetype.hpp at the phases?

I tried this out but could not find a good way to interleave code comments and `flags` entries (only using multi-line comments with additional backslashes, which looks too convoluted in my opinion).

**Comment by chhagedorn** [comment]
> > Just an idea, since you've provided a nice description for each phase in the PR description, should we add these in phasetype.hpp at the phases?
> 
> I tried this out but could not find a good way to interleave code comments and `flags` entries (only using multi-line comments with additional backslashes, which looks too convoluted in my opinion).

I see, that does not seem to be straight forward. I guess then it's okay to omit these descriptions.

**Review by chhagedorn — APPROVED** [review]

**Comment by robcasloz** [comment]
Thanks Daniel and Christian for reviewing!

File: src/hotspot/share/opto/phasetype.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dafedafe) [reply src/hotspot/share/opto/phasetype.hpp]: To be consistent I guess the same could be done for `MERGE_MULTIDEFS`

**Review by abdelhak-zaaim — APPROVED** [review]

**Review by dlunde — APPROVED** [review]

File: src/hotspot/share/opto/phasetype.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/hotspot/share/opto/phasetype.hpp]: Thanks Damon, done in commit fb35674f.

**Comment by robcasloz** [comment]
> Very very cool!

Thanks!

> @robcasloz do you think it could make sense to add a few IR tests just to make sure that the new steps are actually dumped?

I think that would be a good idea, but since we do not have any such test yet, I propose to address that in a separate RFE, starting from the most important phase dumps (i.e. in increasing graph dump level). Does that make sense @dafedafe?

**Comment by dafedafe** [comment]
> I think that would be a good idea, but since we do not have any such test yet, I propose to address that in a separate RFE, starting from the most important phase dumps (i.e. in increasing graph dump level). Does that make sense @dafedafe?

Totally! Thanks @robcasloz.

**Review by dafedafe — APPROVED** [review]

**Comment by robcasloz** [comment]
Thanks Daniel, Christian, and Damon for reviewing! @chhagedorn may I get a re-approval?

**Review by chhagedorn — APPROVED** [review]
Still good!

**Comment by robcasloz** [comment]
Thanks Christian!

**Pull request closed without merging**